Add ability to close tabs and get all responses from navigate

mbutki opened the pull request:
This adds the ability to:
1) Close tabs
2) Get response bodies from a navigate

To ensure we get all responses, if waitForLoad === true, then we:
1) Wait for both the onload event 
2) Count all requests and responses and make sure the two counts match to ensure there are no responses still inflight.

Commit:
Add ability to close tabs and get all responses from navigate

asakusuma commented on src/app-env.ts:
```diff
@@ -98,14 +98,15 @@ export class TestEnvironment<S extends TestServerApi = TestServerApi> {
-  public async closeTab() {
-    // TODO: Call ensureNoErrors() on the tab
+  public async closeTab(targetId: string) {
+    await this.targetIdToClientEnv[targetId].close();
```
Can you check to make sure the `targetId` exists in `this.targetIdToClientEnv`? And then throw a nice descriptive error if it doesn't exist.

mbutki replied:
Good point. I wasn't thinking about this from the independent method call point of view.

asakusuma commented on src/models/frame.ts:
```diff
@@ -39,57 +70,55 @@ export class FrameNavigation {
-export interface PageNavigateResult {
-  networkResult: Network.ResponseReceivedParameters;
-  frame: Page.Frame;
+export interface ResponseHash {
```
Instead of returning a `ResponseHash` from `navigate`, why not keep returning `PageNavigateResult` and then change `PageNavigateResult` to be something like:
```TypeScript
export interface PageNavigateResult {
   page: CapturedResponse;
   childResponses: ResponseHash;
   frame: Page.Frame;
}
```
This also eliminates the need to expose the acceptance utils.

mbutki replied:
I agree that my ResponseHash ended up being a little too generic.

asakusuma reviewed (commented):
Still need to take another pass, but here's some feedback for now

asakusuma commented on src/models/frame.ts:
```diff
@@ -39,57 +70,55 @@ export class FrameNavigation {
       throw new Error('Received network response without frameId');
     }
   }
+  onRequestWillBeSent(req: Network.RequestWillBeSentParameters) {
```
Great idea

asakusuma commented on src/models/client.ts:
```diff
@@ -209,28 +217,37 @@ export class ClientEnvironment {
-  public async navigate(arg?: string | NavigateOptions): Promise<NavigateResult> {
+  private async convertToHash(responses: Network.ResponseReceivedParameters[], shouldAddBodies: boolean) {
+    const resHash = {} as ResponseHash;
+    const bodyExtractions = responses.map(async (response) => {
```
I think map reduce might make this a little easier to read.

```
return responses.map(() => {
   ...
   return capturedResponse;
}).reduce((resHash: ResponseHash, capturedResponse) => {
   resHash[capturedResponse.url] = capturedResponse;
}, {});

asakusuma replied:
I believe this requires adding `url` to `CapturedResponse`, which might be good anyways.

mbutki replied:
The issue here is that "getResponseBody()" is an async operation, but I don't want to just store that promise in resHash[url].body. I want to store the actual resolved body in resHash[url].body. This is because chrome clears away the response body buffer on the next navigation, so want to be storing hard copies of the bodies to be durable across navigates.

If this was not my requirement, I agree that a .map().reduce() would be cleaner. But as it stands, I haven't been able to figure out how to transform my async map into a .map().reduce() which functions the same way.

Let me know if there is a way to achieve the same result with .map().reduce().

mbutki replied:
I had tried something like this before, but wasn't sure if using reduce() to fill out a hashtable was an acceptable usage. I'll rewrite it with map()/reduce(), but over multiple lines distinct to make it easier to digest.

asakusuma commented on src/models/frame.ts:
```diff
@@ -39,57 +70,55 @@ export class FrameNavigation {
+        nav.onRequestWillBeSent(req);
+      }
+    } else {
+      throw new Error('Received network request will be sent without frameId');
```
This error message makes it sound like the issue is a recoverable warning, when we are doing a hard throw in reality

mbutki replied:
I've reworded it to 'Received "request will be sent" event without frameId'.

It is already classified as an error by the use of Error(). I think that should be enough? I thought it redundant to say Error("Error: Received....

asakusuma commented on src/app-env.ts:
```diff
@@ -98,14 +98,19 @@ export class TestEnvironment<S extends TestServerApi = TestServerApi> {
+  public async closeTab(targetId: string) {
+    if (!(targetId in this.targetIdToClientEnv)) {
+      throw new Error(`Can't close tab with targetId:{$targetId}. ` +
+                      'This targetId was not created within this TestEnvironment');
```
nit: technically we don't know that it wasn't created, we just know it no longer exists.

mbutki replied:
I'm assuming there is no other way to close a tab in our system. Just in case, I've made it more ambiguous with "This targetId is not associated with this TestEnvironment"

asakusuma commented on src/models/client.ts:
```diff
@@ -209,30 +219,65 @@ export class ClientEnvironment {
-    const tree = await this.page.getFrameTree();
-    const frameId = tree.frameTree.frame.id;
+  private async buildCapturedResponse(response: Network.ResponseReceivedParameters, shouldAddBodies: boolean) {
+    const capturedResponse = {responseMeta: response.response} as CapturedResponse;
```
Why does this need a cast? Does this not work?
```
const capturedResponse: CapturedResponse = { responseMeta: response.response };
```

mbutki replied:
Yeah, "const capturedResponse: CapturedResponse" works fine. I'm just so unused to having to type in an assignment of a new object in ts that I fell back to the cast syntax by default.

asakusuma commented on src/models/client.ts:
```diff
@@ -209,30 +219,65 @@ export class ClientEnvironment {
 
-    const navPromise = this.frameStore.start(frameId, waitForLoad);
-    await this.page.navigate({ url });
+    if (shouldAddBodies) {
```
should `shouldAddBodies` be renamed to `shouldAddBody`, since there appears to only be one body?

mbutki replied:
Yeah, I've renamed. It was a carry over from the var name in its parent method.

asakusuma commented on src/models/client.ts:
```diff
@@ -209,30 +219,65 @@ export class ClientEnvironment {
-    await this.page.navigate({ url });
+    if (shouldAddBodies) {
+      const body = await this.network.getResponseBody({requestId: response.requestId});
+      capturedResponse['body'] = body.body;
```
I think you can condense this into
```
const body = shouldAddBody ? await this.network.getResponseBody({requestId: response.requestId}) : undefined;
return { responseMeta: response.response, body }
```

mbutki replied:
Interesting. I had assumed that optional items in a TS interface had to either be filled out, or the property excluded. I didn't realize `body = undefined` was an option.

```
export interface CapturedResponse {
  responseMeta: Network.Response;
  body?: string;
}
```

In the util guard for `body` later on, I'm already checking `body === undefined` rather then `body in response` so I guess it's already compatible with this change.

asakusuma commented on src/models/frame.ts:
```diff
@@ -39,57 +70,61 @@ export class FrameNavigation {
- * Represents the result of a navigation
- * @public
- */
+export interface ResponseHash {
```
For naming consistency, might be good to name this `CapturedResponseHash`

mbutki replied:
Done, but be careful letting this old Java developer make long variable/method names ^_^

asakusuma commented on test/acceptance/utils.ts:
```diff
@@ -0,0 +1,14 @@
+  });
+}
+
+export function getBody(response: CapturedResponse) {
```
Is this util necessary? Why do we have this guard for `body`, but not `responseMeta`?

mbutki replied:
responseMeta is required, but body is optional. Bodies are only available if the dev used waitForLoad = true, so trying to get a body when it doesn't exist should be an error case (devs won't try to access just to see if it's there).

```
export interface CapturedResponse {
  responseMeta: Network.Response;
  body?: string;
}
```

asakusuma replied:
I think we can provide safety here without having to add an extra utility. We can effectively have a conditional return type, based off of the parameter, by using overloaded param types.

For example: https://tinyurl.com/y96qpqmy

Alternatively, we can just go ahead and have two separate methods for `navigate`

bottom line is, it would be nicer to be able to eliminate possibility of error, rather than having possibility of runtime error.

mbutki replied:
That's awesome.  I had considered either a dual return type (still with run-time guards), or bifurcating the method (with glue logic to share code). I had no idea TS had method signature overriding. This is the exact solution I needed.

asakusuma replied:
Yea it's a nice feature. I believe it should be used somewhat sparingly though, as it can definitely make the function too complex. In most cases it's probably better to have two different functions, but I'm ok with using the conditional return type here.

mbutki replied:
I think we should just drop the waitForLoad===false code path all together. Below is why:

--- TS method signature overriding ---
I tried it out, but it turns out to be too much of a hack to get working in this more complex case. The issue is that you literally have to pass true/false, not just a boolean variable. This work fine for the first call, where the user might literally call myMethod(true), but in my case the type we want to split on is three functions deep, which mean you'd have to start calling all the nested child functions like this:

if (waitForLoad) {
  childFunction(args, true);
else {
  childFunction (args, false);
} 

as well as have the 3 type signatures for each child function.

--- Split into two functions ---
The code isn't as elegant. I had to make a base function and three interfaces for the "wait=false" call, and another three interfaces & 30 lines of code to take the result obj apart, get bodies, and then jam it all back together.

--- With Either Approach ---
We already have 2 util functions in ember-li-sw which:
A) Are used in almost every test (both wait=true & wait=false tests)
B) Wrap navigate()
Since we now have two navigate functions, I would have to now split those two util functions into four util functions in order to maintain having only one type returned. You can imagine how this won't scale for the future if we add more util functions like this. I could use the "TS method signature overriding" trick to fork off into only one of the two nav functions and maintain only one instance of a util function, but at this point the code base is getting pretty complex.

All of this seem like way too much complexity in the code to support a "return after only nav complete" function that we don't need right now. For our test app (in both ember-li-sw and swae), turning all to waitForLoad doesn't significantly change the total test run time (delta sits within noise variation).

asakusuma replied:
My main issue with `getBody` is that the only value the util is providing is a runtime check, which could be a static compile time check.

> we now have two navigate functions

What are these two functions? `navigate` and ...?

> --- Split into two functions ---
> The code isn't as elegant. I had to make a base function and three interfaces for the ? 
"wait=false" call, and another three interfaces & 30 lines of code to take the result obj apart, get bodies, and then jam it all back together.

Is that all really necessary? Why can't we just rename current `navigate()` to `private loadPage()` and then create `load()` and `navigate()` where the former passes `waitForLoad=true`

> All of this seem like way too much complexity in the code to support a "return after only nav complete" function that we don't need right now. For our test app (in both ember-li-sw and swae), turning all to waitForLoad doesn't significantly change the total test run time (delta sits within noise variation).

So I don't yet buy that it's really that complicated to support both cases. Did you run your perf tests on CI tooling? If it really doesn't make any test perf difference, I'm fine dropping support for short circuiting waiting for the body.

asakusuma commented:
Final merge should probably be just a single squashed commit

mbutki commented:
I had already rewritten the code to have split function before my last post. I've updated this PR with a commit adding the split function if you want to take a look at it. The changes in this repo are not that bad, but the annoying part is having to have two versions of every utility function we make out of these. For example, in ember-li-sw `loadAppInNewClient()` calls `navigateAndWaitForActiveSw()` which calls `navigate()`. To support `loadAppInNewClient()` being about to do a waiting load and a non-waiting load, now we have to split it and all it's children (navigateAndWaitForActiveSw) into two.

> What are these two functions? navigate and ...?
I called them navigate and navigateAndWait. They just call the privateNavigate function with waitForLoad = true/false

> Is that all really necessary?
That's what you get by doing the approach that you mention.

> Did you run your perf tests on CI tooling?
I just used the `time` command on a few runs of both repo's acceptance tests. The variation between runs was large enough that I couldn't make out any perceptible difference between waiting and not waiting. I could automate it, run it through a wilcox test, etc, but that seemed overboard.

mbutki commented:
We have decided to keep load() / navigate() as two separate functions. Shared util function in ember-li-sw will only use the load() version to keep things simple, while more complicated products (e.g. Neptune) can opt for either.

I've cleaner up the code. The return types became bloated over several iterations; this latest commit simplifies them to:
```
export interface PageLoadResult {
  page: CapturedResponse;
  childResponses: CapturedResponseHash;
  frame: Page.Frame;
}
```
for load(), and
```
export interface PageNavigateResult {
  page: CapturedResponse;
  frame: Page.Frame;
}
```
for navigate

asakusuma commented on src/models/client.ts:
```diff
@@ -205,32 +231,60 @@ export class ClientEnvironment {
+  }
+
+  /**
+   * Load page in active tab. Resolves on HTML load. To wait for a full page load, and
```
I think we should specify that this method waits for the HTTP response. I think this is more explicit than "HTML load". I also think we should specifically callout that "full page load" means waiting for the `load` event.

mbutki replied:
cleaned up

asakusuma replied:
Comments still say "HTML load". navigate doesn't care what kind of response it gets, it doesn't have to be HTML. Also I think HTTP response is more specific and accurate.

mbutki replied:
That's weird, I see the latest, squashed commit as containing:
  
/**
   * Load page in active tab. Resolves on HTML response. To wait for a page's load event, and
   * also recieve all asset response data, use load() method instead.
   * @return HTML response and page frame data
   */
  public async navigate(arg?: string | NavigateOptions): Promise<PageNavigateResult> {

asakusuma replied:
Right. Latest commit says "HTML response." I think we should specify "HTTP response" instead.

mbutki replied:
Ha, too many similar acronyms ^_^. All I could read was HTML. 

I've changed all references of `HTML response` to `initial HTTP response`. I tried just `HTTP response`, but `Resolves on HTTP response` sounded a little vague as to which HTTP response it was referring to.

asakusuma commented on src/models/client.ts:
```diff
@@ -205,32 +231,60 @@ export class ClientEnvironment {
+  }
+
+  /**
+   * Load page in active tab. Resolves on page load complete, i.e. Max(onLoad, all responses recieved)
```
I think we should specifically callout that "page load complete" means waiting for the `load` event.

mbutki replied:
I'll expand out the existing `i.e. Max(onLoad, all responses received` into a full sentence. Also you are correct that it is `load` event, not the `onLoad` event. `onLoad` is technically a method which is overridable if you want to execute code on the `load` event. It's so common to swap the two that I had never considered the difference before.

asakusuma reviewed (changes requested):
Some minor documentation issues, and needs to be squashed into a single commit, then I am good to merge. Thanks!
